src/query/api/v1/middleware/rewrite.go

```diff
@@ -117,34 +131,33 @@ func rewriteRangeDuration(
 	if err != nil {
 		return err
 	}
-
-	attrs, err := store.QueryStorageMetadataAttributes(ctx, params.start, params.end, fetchOpts)
+	// Get the appropriate time range before updating the lookback
+	// This is necessary to cover things like the offset and `@` modifiers.
+	startTime, endTime := getQueryBounds(opts, params, fetchOpts, logger)
+	res, err := findLargestQueryResolution(ctx, store, fetchOpts, startTime, endTime)
 	if err != nil {
 		return err
 	}
-
-	// Find the largest resolution
-	var res time.Duration
-	for _, attr := range attrs {
-		if attr.Resolution > res {
-			res = attr.Resolution
+	updateLookback, updatedLookback := maybeUpdateLookback(params, res, opts)
```

Comment:
nit: I would flip these return values. It's more idiomatic Go to return the new value followed by a boolean related to your condition